### PR TITLE
Use FAISS with RMM

### PR DESCRIPTION
**Pull request by viclafargue**
Answers https://github.com/rapidsai/cuml/issues/2821

### CODE REVIEW

File: cpp/include/raft/spatial/knn/faiss_mr.hpp [hunk 1]
```diff
@@ -0,0 +1,333 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
```
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/faiss_mr.hpp:7]: Thanks for opening the issue w/ the FAISS project: https://github.com/facebookresearch/faiss/issues/2097. 

We also need to reference that issue here with a comment that explains why we've temporarily copied this code and how we can remove the copied code once that issue is resolved.

**Review by cjnolet — CHANGES_REQUESTED** [review]
The changes are looking much better. I also want to make sure we clearly describe the path to removing the copied code and adapting it to their API once that issue is resolved. Once that's done, this PR will be ready to go.

**Comment by viclafargue** [comment]
Opened an issue on FAISS to avoid code duplication : facebookresearch/faiss#2097.

Using a custom memory manager in FAISS currently requires overriding methods of `GpuResources` and `GpuResourcesProvider` classes. Most of the logic however, is consistent with existing `StandardGpuResourcesImpl` and `StandardGpuResources` classes causing unnecessary and difficulty maintainable rewrites.

It would be possible to prevent consequent rewrites thanks to updates to the `StandardGpuResources.h` header file. Indeed the ideal situation would be to inherit the logic of these classes and only override the `allocMemory` and `deallocMemory` methods.

The necessary changes would be the following :
1. Promote attributes from private to protected
2. Use `allocMemory` in `initializeForDevice` to prevent `cudaHostAlloc` call
3. (optionally) Remove unnamed namespace to allow the reuse of the `allocsToString` function

**Comment by cjnolet** [comment]
@viclafargue, I looked back over my last comment and realized it was fairly ambiguous. I think having the comment on this PR is very useful as well but we should also have the comment in the code, preferably at the top of `faiss_mr.hpp`

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@viclafargue it's been a little while so just wanted to check the status on this PR. IIRC, we are waiting on cuml side to pass successfuly, right?

**Comment by viclafargue** [comment]
> @viclafargue it's been a little while so just wanted to check the status on this PR. IIRC, we are waiting on cuml side to pass successfuly, right?

Thanks for reminding me of this. Yes, I just solved the merge conflicts.

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**